src/unread/unreadModel.js

```diff
@@ -44,53 +44,153 @@ export const getUnreadMentions = (state: GlobalState): UnreadMentionsState => st
-  });
-}
-
-// Like `updateAll`, but prune values equal to `zero` given by `updater`.
-function updateAllAndPrune<K, V>(
+// Like `Immutable.Map#update`, but prune returned values equal to `zero`.
+function updateAndPrune<K, V>(
   map: Immutable.Map<K, V>,
   zero: V,
-  updater: V => V,
+  key: K,
+  updater: (V | void) => V,
 ): Immutable.Map<K, V> {
-  return map.withMutations(mapMut => {
-    map.forEach((value, key) => {
-      const newValue = updater(value);
-      if (newValue === value) {
-        return; // i.e., continue
-      }
-      if (newValue === zero) {
-        mapMut.delete(key);
-      } else {
-        mapMut.set(key, newValue);
-      }
-    });
-  });
+  const value = map.get(key);
+  const newValue = updater(value);
+  if (newValue === value) {
+    return map;
+  }
+  if (newValue === zero) {
+    return map.delete(key);
+  } else {
+    return map.set(key, newValue);
+  }
+}
+
+/**
+ * Remove the given values from the list.
+ *
+ * This is equivalent to
+ *   list_.filter(x => toDelete.indexOf(x) < 0)
+ * but more efficient.
+ *
+ * Specifically, for n items in the list and k to delete, this takes time
+ * O(n log n) in the worst case.
+ *
+ * In the case where the items to delete all appear at the beginning of the
+ * list, and in the same order, it takes time O(k log n).  (This is the
+ * common case when marking messages as read, which motivates this
+ * optimization.)
+ */
+// In principle this should be doable in time O(k + log n) in the
+// all-at-start case.  We'd need the iterator on Immutable.List to support
+// iterating through the first k elements in O(k + log n) time.  It seems
+// like it should be able to do that, but the current implementation (as of
+// Immutable 4.0.0-rc.12) takes time O(k log n): each step of the iterator
+// passes through a stack of log(n) helper functions.  Ah well.
+//
+// The logs are base 32, so in practice our log(n) is never more than 3
+// (which would be enough for 32**3 = 32768 items), usually at most 2
+// (enough for 1024 items); and for the messages in one conversation, very
+// commonly 1, i.e. there are commonly just ≤32 messages.  So the difference
+// between O(k log n) and O(k + log n) might be noticeable but is unlikely
+// to be catastrophic.
+function deleteFromList<V>(
+  list_: Immutable.List<V>,
+  toDelete_: Immutable.List<V>,
+): Immutable.List<V> {
+  // Alias the parameters because Flow doesn't accept mutating them.
+  let list = list_;
+  let toDelete = toDelete_;
+
+  // First, see if some items to delete happen to be at the start, and
+  // remove those.  This is the common case for marking messages as read,
+  // so it's worth some effort to optimize.  And we can do it efficiently:
+  // for deleting the first k out of n messages, we take time O(k log n)
+  // rather than O(n).
+
+  const minSize = Math.min(list.size, toDelete.size);
+  let i = 0;
+  for (; i < minSize; i++) {
+    // This loop takes time O(log n) per iteration, O(k log n) total.
```

Comment:
Hmm -- it is that, but that makes it also true that it's O(k log n)... and for that matter it's O(n log n), and O(n^2), and O(2^n) and anything else that's bigger :slightly_smiling_face: The [definition](https://en.wikipedia.org/wiki/Big_O_notation) of the "f = O(g)" notation is such that it behaves a lot like a form of less-than-or-equal-to, not like equal-to. It's a bit of a confusing choice of symbols for that semantics, but a very handy semantics.

One reason that semantics is handy is that it lets you state an upper bound, and have it be true even if there's some more-precise upper bound that would also be true. Here, it'd be kind of pathological for k to be more than n -- it would mean we're being informed of some messages marked as read when we didn't know about them as unreads in the first place. The code's behavior needs to be correct in that case (the case can happen if e.g. the user has >50k unreads and goes and catches up on some of the older of them), and also our analysis should be truthful about that case, but it's fine for the analysis to not state the tightest possible bounds about it.

Similarly, for the overall bound lower down:
```
 * existing messages is much larger.  Specifically the time spent should be
 * O(N' log n + c log C), where the messages to delete appear in c out of a
 * total of C conversations, and the affected conversations have a total of
 * N' messages and at most n in any one conversation.  If the messages to be
```
the "at most n in any one conversation" thing probably isn't the most precise bound that'd be possible to give here, but it is accurate, and a more precise one would I think be more complicated to state (involving like the distribution of sizes of different conversations), so this is a convenient bound for thinking about it.

Reply:
Ah, right! Thanks!